### PR TITLE
Add Pipeline Flag and Enable Inference With Pretrained Weights

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
1. Added pipeline flag and implemented for zero_shot_image_classification and feature_extraction based on offline discussion.
2. Added softmax regression loss.
3. Added example for PAWS-X.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/constants.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 MULTICLASS = "multiclass"
 REGRESSION = "regression"
 ZERO_SHOT = "zero_shot"
+ZERO_SHOT_SENTENCE_SIMILARITY = "zero_shot_sentence_similarity"
+ZERO_SHOT_ALL = [ZERO_SHOT, ZERO_SHOT_SENTENCE_SIMILARITY]
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/constants.py:15]: If we go fine-grained, take image classification as an example, we will have `image_classification` and `zero_shot_image_classification`.

For `zero_shot_image_classification`, it will be mainly used for CLIP, LIT kind of models to do inference without training. 

For `image_classification`, it will have three use cases: direct inference, finetuning, and training.

Each use case is different.

(1) zero_shot_image_classification: require both image and text input, no model change, no training, no training data needed, pretrained=True

(2) direct inference: require image as input, no model change, no training, no training data needed, pretrained=True

(3) finetuning: require image as input, has model change (new head), has training and need training data, pretrained_backbone=True

(4) training: require image as input, has model change (new head), has training and need training data, pretrained_backbone=True or you can train from scratch

Other CV tasks are similar. I'm not sure about other applications, but we can discuss problem_types, zero_shot flag, pretrained flag based on these use cases first.

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/constants.py:15]: For text part, we also have a similar scenario like in (1) when the model structure is two tower. Currently user have to extract the 2 embeddings of paired input and then do cosine similarity themselves (like the example in this PR).  I think we may distinguish (1) and (2) by using predict() in (1) and extract_feature() in (2), while the flag zero_shot distinguishes (1)(2) and (3)(4).

**Comment by sxjscience** [comment]
@cheungdaven The PR that adds zero_shot for sentence similarity feature.

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py [hunk 2]
```diff
@@ -246,6 +256,29 @@ def _compute_rkd_loss(
 
         return loss
 
+    def _compute_softmax_regression_loss(
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/lit_distiller.py:259]: Is there a paper for this loss type? It's kind of weird why distillation needs to compute loss in this way.

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py [hunk 3]
```diff
@@ -259,11 +292,19 @@ def _compute_loss(
...
-        loss += soft_label_loss * self.soft_label_weight
+        if self.soft_label_weight > 0:
+            soft_label_loss = self._compute_soft_label_loss(
+                student_output=student_output,
+                teacher_output=teacher_output,
+            )
+            loss += soft_label_loss * self.soft_label_weight
+
+        if self.softmax_regression_weight > 0:
+            softmax_regression_loss = self._compute_softmax_regression_loss(
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/lit_distiller.py:303]: We may need a better name. This loss also belongs to the soft label loss. It just reuses teacher's head.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 4]
```diff
@@ -1431,6 +1464,8 @@ def evaluate(
         data: Union[pd.DataFrame, dict, list],
         metrics: Optional[Union[str, List[str]]] = None,
         return_pred: Optional[bool] = False,
+        label: Optional[str] = None,
+        eval_metric: Optional[str] = None,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1468]: These two new arguments seem not suitable for  `evaluate()`. We need to reconsider the design.

File: multimodal/src/autogluon/multimodal/presets.py [hunk 5]
```diff
@@ -51,14 +51,34 @@ def multilingual():
 
 
 @automm_presets.register()
-def zero_shot():
+def zero_shot_classification():
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/presets.py:54]: Does it make sense to use `zero_shot_classification ` only for text?

File: multimodal/src/autogluon/multimodal/presets.py [hunk 6]
```diff
@@ -51,14 +51,34 @@ def multilingual():
...
+    return {
+        "model.names": ["clip"],
         "model.clip.checkpoint_name": "openai/clip-vit-large-patch14-336",
         "model.clip.max_text_len": 0,
         "env.eval_batch_size_ratio": 1,
     }
 
 
+@automm_presets.register()
+def feature_extraction():
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/presets.py:73]: Ditto. Is `feature_extraction` only for text?

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 7]
```diff
@@ -1450,6 +1485,21 @@ def evaluate(
...
+                    "rmse",
+                    "r2",
+                    "pearsonr",
+                    "spearmanr",
+                ]:
+                    self._problem_type = REGRESSION
+            self._label_column = label
+            self._eval_metric_name = eval_metric
+            self.fit(data, data, init_only=True)
+
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:1502]: Using the code in init and calling fit here looks like a hack. We need a better design.

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py (hunk 2 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/lit_distiller.py:259]: Yes it's an iclr'21 paper: https://www.adrianbulat.com/downloads/ICLR2021/knowledge_distillation_via_softmax_regression_representation_learning.pdf

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py (hunk 3 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/lit_distiller.py:303]: This name is used in the paper, but of course we can change it with a better name.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 4 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:1468]: I'll remove changes for evaluate and we can work on it in next pr.

File: multimodal/src/autogluon/multimodal/presets.py (hunk 5 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/presets.py:54]: In hf this pipeline is NLP only.

File: multimodal/src/autogluon/multimodal/presets.py (hunk 6 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/presets.py:73]: It's should be able to apply to any models.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 7 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:1502]: Yes, I'll remove changes for evaluate and we can work on it in next pr.

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py [hunk 8]
```diff
@@ -246,6 +256,29 @@ def _compute_rkd_loss(
...
+        teacher_output: dict,
+    ):
+        student_feature = student_output[self.student_model.prefix][FEATURES].squeeze(dim=1)
+        student_feature = self.output_feature_adaptor(student_feature)
+
+        student_logits = self.teacher_model.head(student_feature)
+        soft_labels = teacher_output[self.teacher_model.prefix][LOGITS].squeeze(dim=1)
+
+        student_logits = student_logits / self.temperature
+        soft_labels = soft_labels / self.temperature
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/lit_distiller.py:271]: The paper doesn't use temperature for mse.

File: multimodal/src/autogluon/multimodal/configs/distiller/default.yaml [hunk 9]
```diff
@@ -1,9 +1,11 @@
 distiller:
   soft_label_loss_type: ""
+  softmax_regression_loss_type: "mse"
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/distiller/default.yaml:3]: Better to add a paper reference here.

File: multimodal/src/autogluon/multimodal/presets.py (hunk 6 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/presets.py:73]: But setting `"model.names": ["hf_text"]` only allows using the hf text models.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 10]
```diff
@@ -779,6 +804,7 @@ def _fit(
         hyperparameters: Optional[Union[str, Dict, List[str]]] = None,
         teacher_predictor: Union[str, MultiModalPredictor] = None,
         hpo_mode: bool = False,
+        init_only: bool = False,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:807]: Do we need `init_only` for now?

File: multimodal/src/autogluon/multimodal/presets.py (hunk 6 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/presets.py:73]: Yes, this preset is for text only. Or do we need to change it to clip to support both text and image?

File: multimodal/src/autogluon/multimodal/optimization/lit_distiller.py (hunk 8 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/lit_distiller.py]: I think it's similar to soft logit loss so added a temperature here. I just removed it and will report the performance change in readme later.

File: multimodal/src/autogluon/multimodal/configs/distiller/default.yaml (hunk 9 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/configs/distiller/default.yaml:3]: Done

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 10 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py]: Removed. Let's work on to support evaluate before fit in next pr.

File: multimodal/src/autogluon/multimodal/constants.py [hunk 11]
```diff
@@ -10,7 +10,27 @@
...
+QUESTION_ANSWERING = "question_answering"  # TODO
+SUMMARIZATION = "summarization"  # TODO
+TABLE_QUESTION_ANSWERING = "table_question_answering"  # TODO
+TEXT_CLASSIFICATION = "text_classification"  # TODO
+TEXT_GENERATION = "text_generation"  # TODO
+TEXT_TO_TEXT_GENERATION = "text_to_text_generation"  # TODO
+TOKEN_CLASSIFICATION = "token_classification"  # TODO
+TRANSLATION = "translation"  # TODO
+VISUAL_QUESTION_ANSWERING = "visual_question_answering"  # TODO
+ZERO_SHOT_CLASSIFICATION = "zero_shot_classification"  # TODO
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/constants.py:32]: Let's remove the TODOs.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/constants.py:32]: Remove the constants that are marked with `# TODO`

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/constants.py]: Done.

File: examples/automm/pipeline/feature_extraction_example.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    ### Dataset Loading
+    train_df = load_dataset("wietsedv/stsbenchmark", split="train").to_pandas()
+    val_df = load_dataset("wietsedv/stsbenchmark", split="validation").to_pandas()
+    test_df = load_dataset("wietsedv/stsbenchmark", split="test").to_pandas()
+
+    predictor = MultiModalPredictor(pipeline="feature_extraction")
+
+    eval(predictor, test_df)
+
+    # TODO
```
Comment (bryanyzhu) [inline examples/automm/pipeline/feature_extraction_example.py:39]: What is the TODO here?

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 13]
```diff
@@ -140,6 +142,9 @@ def __init__(
             (options: 'binary', 'multiclass', 'regression').
             If `problem_type = None`, the prediction problem type is inferred
             based on the label-values in provided dataset.
+        pipeline
+            This defines inference tasks like FeatureExtraction, ZeroShotClassification, etc.
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/predictor.py:146]: Are these all inference tasks? Or there are training tasks?

File: examples/automm/pipeline/feature_extraction_example.py (hunk 12 above)
Comment (FANGAreNotGnu) [reply examples/automm/pipeline/feature_extraction_example.py:39]: We are not able to run fit for two tower models yet. Thus we may need to wait for that to work on fit after predict.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 13 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:146]: Currently we only use this for inference as what huggingface does: (https://huggingface.co/docs/transformers/main_classes/pipelines). But if later we find this flag useful in training we may change this.

File: examples/automm/pipeline/feature_extraction_example.py (hunk 12 above)
Comment (bryanyzhu) [reply examples/automm/pipeline/feature_extraction_example.py:39]: Got it, then add this message after TODO will be fine, like `TODO: support fit after predict for two tower models.`

Comment (FANGAreNotGnu) [reply examples/automm/pipeline/feature_extraction_example.py]: Sure, just added. Thanks!

**Review by bryanyzhu — APPROVED** [review]

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 14]
```diff
@@ -217,8 +223,10 @@ def __init__(
         self._warn_if_exist = warn_if_exist
         self._enable_progress_bar = enable_progress_bar if enable_progress_bar is not None else True
 
-        if problem_type is not None and problem_type.lower() == ZERO_SHOT:
-            self._config, self._model, self._data_processors = init_zero_shot(hyperparameters=hyperparameters)
+        if pipeline is not None:
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/predictor.py:226]: Should we raise a DeprecatedError about `problem_type == "zero_shot"`? Since it's in 0.5.1.

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:236]: I set the pipeline=ZERO_SHOT_IMAGE_CLASSIFICATION if we see problem_type == "zero_shot" and added a deprecated warning.

**Review by zhiqiangdon — APPROVED** [review]
Thanks!
